### DISCUSSION TITLE
ProductList `VARIANT_MODE_INCLUDE_PARENT_OBJECT`: respect variants in search but only return products

### DISCUSSION THREAD

**Discussion by seybsen** [pimcore/pimcore#10978] [Q&A / Support] [answered]
I built a ProductList like this in my ProductController:

```
$list = $indexService->getProductListForCurrentTenant();
$list->setVariantMode(ProductListInterface::VARIANT_MODE_INCLUDE_PARENT_OBJECT);
...
$listHelper->setupProductList($filterDefinition, $list, $params, $filterService, true);
...
$params['products'] = $list;
...
return $this->renderTemplate('product/list.html.twig', $params);
```

The filtering needs to respect values which are set in variants, but I only want to return the real objects in `$list` without the variants.
Is this possible?

I thought that this would be the default behaviour of `VARIANT_MODE_INCLUDE_PARENT_OBJECT`:
https://github.com/pimcore/pimcore/blob/119f2a1f49c158fad294ab1eb6e697295612d5da/bundles/EcommerceFrameworkBundle/IndexService/ProductList/ProductListInterface.php#L53

Background info:
The detail pages of the products show all variants so I don't need them in the category listing.

**Comment by fashxp** [comment, marked as answer]
Hi, 
yes `VARIANT_MODE_INCLUDE_PARENT_OBJECT` should do exactly what you are looking for.

**Reply by seybsen** [reply]
Hi @fashxp, thanks for your response. Unfortunately it does not work like I thought it would or I'm doing something wrong:

![image](https://user-images.githubusercontent.com/739688/145374868-ae7638ce-5dad-405c-8c0e-49a5cf10c785.png)

I still have products in `$list` of type `variant`, not only those of type `object` like I expected.

Maybe this is because of variants of variants? Like this:
![image](https://user-images.githubusercontent.com/739688/145390331-611356be-b3ad-47d8-840a-eb462a46521e.png)

ID `99` (object) and `104` (variant) are included in `$list` because the `o_virtualProductId` (which is used for `GROUP BY`/`DISTINCT`) is different for variants of variants:

|o_id|o_virtualProductId|o_parentId|o_type |
|----|------------------|----------|-------|
|99  |99                |13        |object |
|104 |99                |99        |variant|
|105 |104               |104       |variant|
|106 |104               |104       |variant|


https://github.com/pimcore/pimcore/blob/4550a053fac47a1d2c0406bf15c9799ef078bbb7/bundles/EcommerceFrameworkBundle/IndexService/ProductList/DefaultMysql/Dao.php#L73-L84

**Reply by fashxp** [reply]
you need to implement the `getOSParentId` method so that it returns the `id` of the main product and not just the next parent id ...
similar to how we do it in [our demo](https://github.com/pimcore/demo/blob/10.x/src/Model/Product/Car.php#L71)

**Reply by seybsen** [reply]
I'll try this. Thank you very much!
